packages/astro/src/env/vite-plugin-env-virtual-mod.ts

```diff
@@ -1,28 +1,117 @@
-import type { Plugin } from 'vite';
+import { loadEnv, type Plugin } from 'vite';
 import type { AstroSettings } from '../@types/astro.js';
 import type { Logger } from '../core/logger/core.js';
+import {
+	ENV_TYPES_FILE,
+	RESOLVED_VIRTUAL_CLIENT_MODULE_ID,
+	VIRTUAL_CLIENT_MODULE_ID,
+} from './constants.js';
+import type { EnvSchema } from './schema.js';
+import { validateEnvVariable } from './validators.js';
+import { AstroError, AstroErrorData } from '../core/errors/index.js';
+import type fsMod from 'node:fs';
+import { fileURLToPath } from 'node:url';
 
 interface AstroEnvVirtualModPluginParams {
 	settings: AstroSettings;
 	logger: Logger;
+	mode: 'dev' | 'build' | string;
+	fs: typeof fsMod;
 }
 
 export function astroEnvVirtualModPlugin({
 	settings,
 	logger,
+	mode,
+	fs,
 }: AstroEnvVirtualModPluginParams): Plugin | undefined {
 	if (!settings.config.experimental.env) {
 		return;
 	}
 
 	logger.warn('env', 'This feature is experimental. TODO:');
 
-	const { schema } = settings.config.experimental.env;
-	// TODO: client / public
+	const schema = settings.config.experimental.env.schema ?? {};
+	const loadedEnv = loadEnv(
+		mode === 'dev' ? 'development' : 'production',
+		fileURLToPath(settings.config.root),
+		''
+	);
+
+	const { clientContent, clientDts } = handleClientModule({ schema, loadedEnv });
+	handleDts({ settings, fs, content: clientDts });
```

Comment:
We are generating code before the `load` function; is that correct? Shouldn't we generate code inside `load`? Types and client code will be emitted only if the user attempts to use the virtual module.

Reply:
Yeah, I also think we should be lazy loading the virtual module code. One thing to also check though is if new env vars get picked up across server restarts. As a Vite plugin, it's best to cache or reset the cache with the `buildStart`/`buildEnd`/`config`/`configResolved` hooks etc so that we re-compute the env when needed.